dependency/consul_exported_services.go

```diff
@@ -0,0 +1,162 @@
+package dependency
+
+import (
+	"log"
+	"net/url"
+	"slices"
+	"time"
+
+	capi "github.com/hashicorp/consul/api"
+	"github.com/pkg/errors"
+)
+
+const (
+	exportedServicesEndpointLabel = "list.exportedServices"
+
+	// ListExportedServicesQuerySleepTime is the amount of time to sleep between
+	// queries, since the endpoint does not support blocking queries.
+	ListExportedServicesQuerySleepTime = 15 * time.Second
+)
+
+// Ensure implements
+var _ Dependency = (*ListExportedServicesQuery)(nil)
+
+// ListExportedServicesQuery is the representation of a requested exported services
+// dependency from inside a template.
+type ListExportedServicesQuery struct {
+	stopCh    chan struct{}
+	partition string
+}
+
+type ExportedService struct {
+	// Name of the service
+	Service string
+
+	// Partition of the service
+	Partition string
+
+	// Namespace of the service
+	Namespace string
+
+	// Consumers is a list of downstream consumers of the service.
+	Consumers ResolvedConsumers
+}
+
+type ResolvedConsumers struct {
+	Peers          []string
+	Partitions     []string
+	SamenessGroups []string
+}
+
+func fromConsulExportedService(svc capi.ExportedService) ExportedService {
+	peers := make([]string, 0, len(svc.Consumers))
+	partitions := make([]string, 0, len(svc.Consumers))
+	samenessGroups := make([]string, 0, len(svc.Consumers))
+	for _, consumer := range svc.Consumers {
+		if consumer.Peer != "" {
+			peers = append(peers, consumer.Peer)
+		}
+		if consumer.Partition != "" {
+			partitions = append(partitions, consumer.Partition)
+		}
+		if consumer.SamenessGroup != "" {
+			samenessGroups = append(samenessGroups, consumer.SamenessGroup)
+		}
+	}
+
+	return ExportedService{
+		Service: svc.Name,
+		Consumers: ResolvedConsumers{
+			Peers:          peers,
+			Partitions:     partitions,
+			SamenessGroups: samenessGroups,
+		},
+	}
+}
+
+// NewListExportedServicesQuery parses a string of the format @dc.
+func NewListExportedServicesQuery(s string) (*ListExportedServicesQuery, error) {
+	return &ListExportedServicesQuery{
+		stopCh:    make(chan struct{}, 1),
```

Comment:
Stop channels do not need to have capacity so long as you only block on `<-stopCh` and signal with `close(stopCh)`.